File: bindings/python/src/graphqlite/algorithms/paths.py
"""Path finding algorithms mixin."""
from __future__ import annotations

from typing import Optional

from ..graph._base import BaseMixin
from ._parsing import extract_algo_array, safe_float, safe_int


class PathsMixin(BaseMixin):
    """Mixin providing path finding algorithm methods."""

    def shortest_path(
        self,
        source_id: str,
        target_id: str,
        weight_property: Optional[str] = None
    ) -> dict:
        """
        Find the shortest path between two nodes using Dijkstra's algorithm.

        Args:
            source_id: Source node's id property value
            target_id: Target node's id property value
            weight_property: Optional edge property to use as weight
                           (if None, uses unweighted/hop count)

        Returns:
            Dict with 'path' (list of node ids), 'distance', and 'found' (bool)
        """
        esc_source = self._escape(source_id)
        esc_target = self._escape(target_id)

        if weight_property:
            esc_weight = self._escape(weight_property)
            query = f'RETURN dijkstra("{esc_source}", "{esc_target}", "{esc_weight}")'
        else:
            query = f'RETURN dijkstra("{esc_source}", "{esc_target}")'

        result = self._conn.cypher(query)

        if len(result) == 0:
            return {"path": [], "distance": None, "found": False}

        row = result[0]

        # Handle nested column_0 structure from algorithm return
        if "column_0" in row:
            data = row["column_0"]
            if isinstance(data, dict):
                return {
                    "path": data.get("path", []),
                    "distance": data.get("distance"),
                    "found": data.get("found", False)
                }

        # Direct access if already unpacked
        return {
            "path": row.get("path", []),
            "distance": row.get("distance"),
            "found": row.get("found", False)
        }

    # Alias
    dijkstra = shortest_path

    def astar(
        self,
        source_id: str,
        target_id: str,
        lat_prop: str | None = None,
        lon_prop: str | None = None
    ) -> dict:
        """
        Find shortest path using A* algorithm with heuristic guidance.

        A* extends Dijkstra's algorithm with a heuristic function that
        estimates distance to the target, making it faster when node
        coordinates are available.

        Args:
            source_id: Starting node's id property value
            target_id: Target node's id property value
            lat_prop: Property name for latitude/y coordinate (optional)
            lon_prop: Property name for longitude/x coordinate (optional)

        Returns:
            Dict with 'path' (list of node ids), 'distance', 'found', 'nodes_explored'
        """
        if lat_prop and lon_prop:
            query = f"RETURN astar('{self._escape(source_id)}', '{self._escape(target_id)}', '{lat_prop}', '{lon_prop}')"
        else:
            query = f"RETURN astar('{self._escape(source_id)}', '{self._escape(target_id)}')"

        result = self._conn.cypher(query)

        if not result:
            return {"path": [], "distance": None, "found": False, "nodes_explored": 0}

        row = result[0]
        return {
            "path": row.get("path", []),
            "distance": row.get("distance"),
            "found": row.get("found", False),
            "nodes_explored": safe_int(row.get("nodes_explored"))
        }

    # Alias
    a_star = astar

    def all_pairs_shortest_path(self) -> list[dict]:
        """
        Compute shortest paths between all pairs of nodes.

        Uses Floyd-Warshall algorithm with O(V³) time complexity and
        O(V²) space complexity. Only practical for graphs with < 10K nodes.

        Returns:
            List of dicts with 'source', 'target', 'distance' for each
            reachable pair of nodes (excludes self-loops and unreachable pairs)
        """
        result = self._conn.cypher("RETURN apsp()")
        rows = extract_algo_array(result.to_list())

        paths = []
        for row in rows:
            source = row.get("source")
            target = row.get("target")
            distance = row.get("distance")

            if source is not None and target is not None:
                paths.append({
                    "source": source,
                    "target": target,
                    "distance": safe_float(distance)
                })

        return paths

    # Alias
    apsp = all_pairs_shortest_path

File: bindings/rust/src/algorithms/paths.rs
//! Path finding algorithm implementations.

use super::parsing::extract_float;
use super::{AStarResult, ApspResult, ShortestPathResult};
use crate::graph::Graph;
use crate::utils::escape_string;
use crate::{Result, Value};

impl Graph {
    /// Find the shortest path between two nodes using Dijkstra's algorithm.
    ///
    /// # Arguments
    ///
    /// * `source_id` - ID of the source node
    /// * `target_id` - ID of the target node
    /// * `weight_property` - Optional edge property to use as weight
    pub fn shortest_path(
        &self,
        source_id: &str,
        target_id: &str,
        weight_property: Option<&str>,
    ) -> Result<ShortestPathResult> {
        let esc_source = escape_string(source_id);
        let esc_target = escape_string(target_id);

        let query = match weight_property {
            Some(wp) => format!(
                "RETURN dijkstra(\"{}\", \"{}\", \"{}\")",
                esc_source,
                esc_target,
                escape_string(wp)
            ),
            None => format!("RETURN dijkstra(\"{}\", \"{}\")", esc_source, esc_target),
        };

        let result = self.connection().cypher(&query)?;

        if result.is_empty() {
            return Ok(ShortestPathResult {
                path: Vec::new(),
                distance: None,
                found: false,
            });
        }

        let row = &result[0];

        // Handle nested column_0 structure
        if let Some(Value::Object(data)) = row.get_value("column_0") {
            let path = data
                .get("path")
                .and_then(|v| match v {
                    Value::Array(arr) => Some(
                        arr.iter()
                            .filter_map(|v| match v {
                                Value::String(s) => Some(s.clone()),
                                _ => None,
                            })
                            .collect(),
                    ),
                    _ => None,
                })
                .unwrap_or_default();

            let distance = data.get("distance").and_then(|v| match v {
                Value::Float(f) => Some(*f),
                Value::Integer(i) => Some(*i as f64),
                _ => None,
            });

            let found = data
                .get("found")
                .and_then(|v| match v {
                    Value::Bool(b) => Some(*b),
                    _ => None,
                })
                .unwrap_or(false);

            return Ok(ShortestPathResult {
                path,
                distance,
                found,
            });
        }

        // Direct access
        let path = row
            .get_value("path")
            .and_then(|v| match v {
                Value::Array(arr) => Some(
                    arr.iter()
                        .filter_map(|v| match v {
                            Value::String(s) => Some(s.clone()),
                            _ => None,
                        })
                        .collect(),
                ),
                _ => None,
            })
            .unwrap_or_default();

        let distance = row.get_value("distance").and_then(|v| match v {
            Value::Float(f) => Some(*f),
            Value::Integer(i) => Some(*i as f64),
            _ => None,
        });

        let found = row
            .get_value("found")
            .and_then(|v| match v {
                Value::Bool(b) => Some(*b),
                _ => None,
            })
            .unwrap_or(false);

        Ok(ShortestPathResult {
            path,
            distance,
            found,
        })
    }

    /// Find shortest path using A* algorithm with heuristic guidance.
    ///
    /// # Arguments
    ///
    /// * `source_id` - Starting node's id
    /// * `target_id` - Target node's id
    /// * `lat_prop` - Optional property name for latitude
    /// * `lon_prop` - Optional property name for longitude
    pub fn astar(
        &self,
        source_id: &str,
        target_id: &str,
        lat_prop: Option<&str>,
        lon_prop: Option<&str>,
    ) -> Result<AStarResult> {
        let esc_source = escape_string(source_id);
        let esc_target = escape_string(target_id);

        let query = match (lat_prop, lon_prop) {
            (Some(lat), Some(lon)) => format!(
                "RETURN astar('{}', '{}', '{}', '{}')",
                esc_source, esc_target, lat, lon
            ),
            _ => format!("RETURN astar('{}', '{}')", esc_source, esc_target),
        };

        let result = self.connection().cypher(&query)?;

        if result.is_empty() {
            return Ok(AStarResult {
                path: Vec::new(),
                distance: None,
                found: false,
                nodes_explored: 0,
            });
        }

        let row = &result[0];

        let path = row
            .get_value("path")
            .and_then(|v| match v {
                Value::Array(arr) => Some(
                    arr.iter()
                        .filter_map(|v| match v {
                            Value::String(s) => Some(s.clone()),
                            _ => None,
                        })
                        .collect(),
                ),
                _ => None,
            })
            .unwrap_or_default();

        let distance = row.get_value("distance").and_then(|v| match v {
            Value::Float(f) => Some(*f),
            Value::Integer(i) => Some(*i as f64),
            _ => None,
        });

        let found = row
            .get_value("found")
            .and_then(|v| match v {
                Value::Bool(b) => Some(*b),
                _ => None,
            })
            .unwrap_or(false);

        let nodes_explored = row
            .get_value("nodes_explored")
            .map(|v| match v {
                Value::Integer(i) => *i,
                _ => 0,
            })
            .unwrap_or(0);

        Ok(AStarResult {
            path,
            distance,
            found,
            nodes_explored,
        })
    }

    /// Compute shortest paths between all pairs of nodes.
    ///
    /// Uses Floyd-Warshall algorithm with O(V³) time complexity.
    pub fn apsp(&self) -> Result<Vec<ApspResult>> {
        let result = self.connection().cypher("RETURN apsp()")?;

        let mut paths = Vec::new();
        for row in result.iter() {
            let source = row.get_value("source").and_then(|v| match v {
                Value::String(s) => Some(s.clone()),
                _ => None,
            });
            let target = row.get_value("target").and_then(|v| match v {
                Value::String(s) => Some(s.clone()),
                _ => None,
            });

            if let (Some(source), Some(target)) = (source, target) {
                paths.push(ApspResult {
                    source,
                    target,
                    distance: extract_float(row, "distance"),
                });
            }
        }
        Ok(paths)
    }
}
